src/jvmMain/kotlin/fr/acinq/lightning/io/JvmTcpSocket.kt

```diff
@@ -58,34 +64,91 @@ class JvmTcpSocket(val socket: Socket) : TcpSocket {
+                        }
+                    }
+                    is TcpSocket.TLS.PINNED_PUBLIC_KEY -> {
+                        logger.info { "using certificate pinning for connections with $host" }
+                        socket.tls(coroutineContext, TLSConfigBuilder().apply {
+                            // build a default X509 trust manager.
+                            val factory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm())!!
+                            factory.init(null as KeyStore?)
+                            val defaultX509TrustManager = factory.trustManagers!!.filterIsInstance<X509TrustManager>().first()
+
+                            // create a new trust manager that always accepts certificates for the pinned public key, or falls back to standard procedure.
                             trustManager = object : X509TrustManager {
-                                override fun checkClientTrusted(p0: Array<out X509Certificate>?, p1: String?) {}
-                                override fun checkServerTrusted(p0: Array<out X509Certificate>?, p1: String?) {}
-                                override fun getAcceptedIssuers(): Array<X509Certificate>? = null
+                                override fun checkClientTrusted(chain: Array<out X509Certificate>?, authType: String?) {
+                                    defaultX509TrustManager.checkClientTrusted(chain, authType)
+                                }
+
+                                override fun checkServerTrusted(chain: Array<out X509Certificate>?, authType: String?) {
+                                    val serverKey = try {
+                                        JvmTcpSocket.buildPublicKey(chain?.asList()?.firstOrNull()?.publicKey?.encoded ?: throw RuntimeException("certificate missing"))
+                                    } catch (e: Exception) {
+                                        logger.error(e) { "failed to read server's pubkey=${tls.pubKey}" }
+                                        throw e
+                                    }
+
+                                    val pinnedKey = try {
+                                        JvmTcpSocket.buildPublicKey(Base64.getDecoder().decode(tls.pubKey))
+                                    } catch (e: Exception) {
+                                        logger.error(e) { "failed to read pinned pubkey=${tls.pubKey}" }
+                                        throw e
+                                    }
+
+                                    if (serverKey == pinnedKey) {
+                                        logger.info { "successfully checked server's certificate against pinned pubkey" }
+                                    } else {
+                                        logger.warning { "server's certificate does not match pinned pubkey, fallback to default check" }
+                                        defaultX509TrustManager.checkServerTrusted(chain, authType)
```

Comment:
@robbiehanson Not sure what we should do in that case. We could reject the connection attempt and dispatch an event containing the new pubkey of the server, and then Phoenix could display a message telling the user that the pubkey changed, and asking whether the app should trust this new key or keep the connection closed. Is that what you had in mind?

Reply:
> We could reject the connection attempt

I think that's the safest option. Because otherwise a malicious actor could simply redirect to a server with a valid certificate.

> and dispatch an event [...], and then Phoenix could display a message

That's a great idea. As of now, the iOS code would simply stop connecting to Electrum. This idea is much better.

Reply:
Done with 357b41536936c55961ebc2b1b82d3f43dfb028d6 and f42bf230c242d639305c19bb03758b948f65e2c0

On the JVM, we throw a `TcpSocket.IOException.ConnectionClosed` exception where `cause` will be a `java.security.cert.CertificateException`.